crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+
+        let events = response
+            .events_before
+            .into_iter()
+            .rev()
+            .chain(response.event)
+            .chain(response.events_after)
+            .collect();
+
+        Ok(StartFromResult { events, has_prev, has_next })
+    }
+
+    /// Runs a backward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the start of the timeline before.
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_backward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Backward).await
+    }
+
+    /// Runs a forward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the end of the timeline before.
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_forward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Forward).await
+    }
+
+    async fn paginate(&self, dir: Direction) -> Result<PaginationResult, PaginatorError> {
+        self.check_state(PaginatorState::Idle)?;
+
+        let token = match dir {
+            Direction::Backward => {
+                let prev_batch_token = self.prev_batch_token.lock().await;
+                if prev_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                prev_batch_token.clone()
+            }
+
+            Direction::Forward => {
+                let next_batch_token = self.next_batch_token.lock().await;
+                if next_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                next_batch_token.clone()
+            }
+        };
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // idle. This check makes sure there's at most one which can set the
+        // state to paginating, preventing a race condition.
+        if self.state.set_if_not_eq(PaginatorState::Paginating).is_none() {
+            return Err(PaginatorError::InvalidPreviousState {
+                expected: PaginatorState::Idle,
+                actual: PaginatorState::Paginating,
+            });
+        }
+
+        let response_result =
+            self.room.messages(MessagesOptions::new(dir).from(token.as_deref())).await;
+
+        // In case of error, reset the state to idle.
+        let response = match response_result {
+            Ok(res) => res,
+            Err(err) => {
+                self.state.set(PaginatorState::Idle);
+                return Err(err);
+            }
+        };
+
+        let hit_end_of_timeline = match dir {
```

Comment:
This structure of 2 methods for forward/backward, then one method for paginate, then `match`es inside that method, makes me feel a bit funny, but I can't see a good way out at the moment, unless you had an enum for the directions that had methods like `hit_end_of_timeline` on it? Not sure it would work, and would be a bit awkward for the early return on line 232.

Reply:
This method does feel slightly long to me, so if we could figure a way to break off into direction-specific submethods, that would be cool...

Reply:
How would you split that up? This method is 60 lines of code long (including 14 empty lines or comment lines), and the duplication in the matches seems less terrible than duplicating the entire methods, so I'm all up for ideas!

Reply:
OK, so I tried it out just to see, and it's not too bad: https://github.com/matrix-org/matrix-rust-sdk/pull/3325/files

Reply:
If `DirThing` had a good name to express what it was, maybe this would be ... good?

Very happy to leave it as it is, just thinking out loud.

Reply:
I don't think introducing 30 new lines and a new data structure instead of keeping the 66-lines long method as is is a net positive, so I'll keep the method as it is right now. Or I can just pass a `&RwLock<Option<String>>` to this `paginate` helper :-)

Reply:
Agree, was fun trying it out :-)